image_view/src/nodes/image_view.cpp

```diff
@@ -32,7 +32,62 @@
 *  POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/
 #include <ros/ros.h>
-#include <nodelet/loader.h>
+#include <image_transport/image_transport.h>
+
+#include <cv_bridge/cv_bridge.h>
+#include <opencv2/highgui/highgui.hpp>
+
+#include <boost/format.hpp>
+#include <boost/thread.hpp>
+
+int g_count;
+bool g_new_msg = false;
+cv::Mat g_last_image;
+boost::format g_filename_format;
+boost::mutex g_image_mutex;
+
+void imageCb(const sensor_msgs::ImageConstPtr& msg)
+{
+  // We want to scale floating point images so that they display nicely
+  bool do_dynamic_scaling = (msg->encoding.find("F") != std::string::npos);
+
+  // Convert to OpenCV native BGR color
+  try {
+    g_last_image = cv_bridge::cvtColorForDisplay(cv_bridge::toCvShare(msg), "",
+                                                 do_dynamic_scaling)->image;
+    g_new_msg = true;
+  } catch (cv_bridge::Exception& e) {
+    ROS_ERROR_THROTTLE(30, "Unable to convert '%s' image for display: '%s'",
+                       msg->encoding.c_str(), e.what());
+  }
+}
+
+static void mouseCb(int event, int x, int y, int flags, void* param)
+{
+  if (event == cv::EVENT_LBUTTONDOWN) {
+    ROS_WARN_ONCE("Left-clicking no longer saves images. Right-click instead.");
+    return;
+  } else if (event != cv::EVENT_RBUTTONDOWN) {
+    return;
+  }
+
+  const cv::Mat &image = g_last_image;
+
+  if (image.empty()) {
+    ROS_WARN("Couldn't save image, no data!");
+    return;
+  }
+
+  std::string filename = (g_filename_format % g_count).str();
+  if (cv::imwrite(filename, image)) {
```

Comment:
dangerous in case the cb thread write to g_last_image: you probably want to protect all that with a mutex (same with imshow).